### ISSUE TITLE
Implicit summing with weirdly nested pmaps

### ISSUE THREAD

**Issue by Thenerdstation** [google/jax#1679]
This code on a TPU:

```python
def f(a, b):
  def g(a, b):
    return b

  def k(q):
    def h(a):
      return jax.pmap(g, "blue")(a, q)
    
    return jax.pmap(h, "green")(a)
    
  return jax.pmap(k, "red")(b)

f(np.ones((2, 2, 2, 2)), np.ones((2, 2, 2, 2)))
```

Prints out this

```
array([[[[[2., 2.],
          [2., 2.]],

         [[2., 2.],
          [2., 2.]]],


        [[[0., 0.],
          [0., 0.]],

         [[0., 0.],
          [0., 0.]]]],



       [[[[2., 2.],
          [2., 2.]],

         [[2., 2.],
          [2., 2.]]],


        [[[0., 0.],
          [0., 0.]],

         [[0., 0.],
          [0., 0.]]]]], dtype=float32)
```

Even though this code basically just returns the "b" array, it somehow is adding parts of the matrix together and zeroing out the rest.

Now, this code is super jank, but I still don't understand where the 2s and 0s are coming from.

**Comment by Thenerdstation** [comment]
jaxpr of `f`

```
{ lambda c ;  ; a b.
  let d = xla_pmap[ axis_name=red
                    axis_size=2
                    devices=None
                    backend=None ] c b
        { lambda  ; c ; b a.
          let d = xla_pmap[ axis_name=green
                            axis_size=2
                            devices=None
                            backend=None ] b c
                { lambda  ; b ; d a.
                  let c = xla_pmap[ axis_name=blue
                                    axis_size=2
                                    devices=None
                                    backend=None ] a b
                        { lambda  ;  ; a b.
                          let 
                          in [b] } [  ;  ]
                      e = gather[ dimension_numbers=GatherDimensionNumbers(offset_dims=(0, 1), collapsed_slice_dims=(0,), start_index_map=(0,))
                                  slice_sizes=(1, 2, 2)
                                  operand_shape=(2, 2, 2) ] c d
                  in [e] } [  ; a ]
          in [d] } [  ; a ]
  in [d] }
```

**Comment by mattjj** [comment]
Thanks for catching this, @Thenerdstation ! It was a bug in our nested `pmap` logic that only struck at triply-nested and higher; it happened to work correctly for just two levels of nesting. We fixed this in #1682.
